src/policy/ancestor_packages.cpp

```diff
@@ -0,0 +1,202 @@
+        if (m_txid_to_entry.count(parent_txid) == 0) continue;
+        auto parent_tx = m_txid_to_entry.at(parent_txid).tx;
+        if (m_txid_to_entry.at(parent_txid).ancestor_subset.empty()) {
+            visit(parent_tx);
+        }
+        auto parent_ancestor_set = m_txid_to_entry.at(parent_txid).ancestor_subset;
+        Assume(!parent_ancestor_set.empty());
+        my_ancestors.insert(parent_ancestor_set.cbegin(), parent_ancestor_set.cend());
+    }
+    entry.ancestor_subset = std::move(my_ancestors);
+}
+
+AncestorPackage::AncestorPackage(const Package& txns_in)
+{
+    // Duplicate transactions are not allowed, as they will result in infinite visit() recursion.
+    Assume(IsConsistent(txns_in));
+    if (txns_in.empty() || !IsConsistent(txns_in)) return;
+    // Populate m_txid_to_entry for quick lookup
+    std::transform(txns_in.cbegin(), txns_in.cend(), std::inserter(m_txid_to_entry, m_txid_to_entry.end()),
+            [](const auto& tx) { return std::make_pair(tx->GetHash(), PackageEntry(tx)); });
+    // DFS-based algorithm to sort transactions by ancestor count and populate ancestor_subset.
+    // Best case runtime is if the package is already sorted and no recursive calls happen.
+    // An empty PackageEntry::ancestor_subset is equivalent to not yet being processed.
+    size_t i{0};
+    while (i < txns_in.size()) {
+        const auto& tx = txns_in[i];
+        if (m_txid_to_entry.at(tx->GetHash()).ancestor_subset.empty()) visit(tx);
+        Assume(!m_txid_to_entry.at(tx->GetHash()).ancestor_subset.empty());
+        ++i;
+    }
+    m_txns = txns_in;
+    // Sort by the number of in-package ancestors.
+    std::sort(m_txns.begin(), m_txns.end(), [&](const CTransactionRef& a, const CTransactionRef& b) -> bool {
+        return m_txid_to_entry.at(a->GetHash()) < m_txid_to_entry.at(b->GetHash());
+    });
+    Assume(IsSorted(m_txns));
+    m_is_ancestor_package = m_txid_to_entry.at(m_txns.back()->GetHash()).ancestor_subset.size() == m_txns.size();
+    // Now populate the descendant caches
+    for (const auto& [txid, entry] : m_txid_to_entry) {
+        for (const auto& anc_txid : entry.ancestor_subset) {
+            m_txid_to_entry.at(anc_txid).descendant_subset.insert(txid);
+        }
+    }
+}
+
+Package AncestorPackage::FilteredTxns() const
+{
+    Package result;
+    for (const auto& [_, entry] : m_txid_to_entry) {
+        if (!entry.skip && !entry.dangles) result.push_back(entry.tx);
+    }
+    std::sort(result.begin(), result.end(), [&](const CTransactionRef& a, const CTransactionRef& b) -> bool {
+        return m_txid_to_entry.at(a->GetHash()) < m_txid_to_entry.at(b->GetHash());
+    });
+    return result;
+}
+
+std::optional<std::vector<CTransactionRef>> AncestorPackage::GetAncestorSet(const CTransactionRef& tx)
+{
+    if (m_txid_to_entry.count(tx->GetHash()) == 0) return std::nullopt;
+    const auto& entry = m_txid_to_entry.find(tx->GetHash())->second;
```

Comment:
In e82f4e7d7b41592b8891a2c6f9b961b2fc2ec51e "[txpackages] add AncestorPackage for linearizing packages"

nit: avoid looking up twice

```suggestion
    const auto& entry_it = m_txid_to_entry.find(tx->GetHash());
    if (entry_it == m_txid_to_entry.end()) return std::nullopt;
    const auto& entry = entry_it->second;
```

Reply:
Done